aws_lambda_builders/workflows/java_gradle/DESIGN.md

```diff
@@ -0,0 +1,117 @@
+# Java - Gradle Lambda Builder
```

Comment:
Got a bigger hypothetical question.

Java build systems (gradle / maven) rely a lot on internal dependencies and sub-projects (sub-modules). How do we want to leverage that?

Example project layout:
```
gradlew
template.yaml
build.gradle
settings.gradle

lambda1/src/main/java
lambda1/build.gradle

lambda2/src/main/java
lambda2/build.gradle

common/src/main/java
common/build.gradle
```

Where template.yaml defines 2 lambda functions:
1. Lambda1 with codeUri: lambda1
2. Lambda1 with codeUri: Lambda2

settings.gradle defines 2 sub-projects:
```
rootProject.name = 'FooService'
include 'lambda1'
include 'lambda2'
```

Lambda1 has a compile dependency on `project(":common")

Should sam build actually search up for the root and then run `./gradlew :Lambda1:build :lambda1:package` to build Lamda1 (and same for Lambda2)

Or do we want to force people to have to go and make 2 full Gradle projects (which is not idiomatic) and foce people to do `common/gradelw build install && sam build` so that common code gets compiled. I would say this is bad, so we may need to tweak the workflow design to allow for this...

(In maven world, it can build just the sub-module's pom.xml)

Reply:
As of this moment, sam build works by by looking at each lambda function/ resource  in isolation and building it.

Look at below code to see how sam build works.

https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/lib/build/app_builder.py#L112

If there are common dependencies, the manifest that is passed in needs to reflect it. The actions in the workflow need to know which dependency to build from where.

Reply:
Hmm okay, the manifest being `build.gradle` in this case. I'm not sure that they *is* a way for it to reflect that, since as @abrooksv pointed out, only the root project `settings.gradle` declares that.

Reply:
Okay actually it may actually not be very difficult. We should be able to leverage the built-in [`getPath`](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#getPath()) API

Reply:
Gradle will handle building other projects that it needs based on the dependency graph (compile project(":common")), the question to SAM is:

How much is mounted in the container? 
1. Is it the entire project? or just the Lambda1 folder?
2. If 1 is everything, do we have a path to it somehow in the workflow?

Reply:
I think there's still some open questions here.

We can only rely on `getProjectDir` if we can run the init script, but that requires a `gradle`/`gradlew` executable. Since we give preference to bundled `gradlew`, we need to be able to find the root directory of the project to search for it.

Is there any way for SAM to give us the root of the project? i.e. if `template.yaml` must be in the root of the project, can SAM provide it to the build function?

Reply:
If it comes to it, we can emulate the root project resolution Gradle does I suppose https://github.com/gradle/gradle/blob/master/subprojects/core/src/test/groovy/org/gradle/initialization/layout/BuildLayoutFactoryTest.groovy

Reply:
Let's not get hung up on what SAM needs or what is mounted. SAM is a caller of this and we should define what is needing without knowing or caring about what SAM does. For example, a customer can use this feature without the docker-lambda build container without ever using SAM CLI at all. We need to think about this in context of what is provided to the [Workflow](https://github.com/awslabs/aws-lambda-builders/blob/develop/aws_lambda_builders/workflow.py#L76)